python/simulation.py

```diff
@@ -1633,22 +1633,36 @@ def _evaluate_dft_objects(self):
             if dft.swigobj is None:
                 dft.swigobj = dft.func(*dft.args)
 
-    def add_dft_fields(self, components, freq_min, freq_max, nfreq, where=None, center=None, size=None, yee_grid=False):
+    def add_dft_fields(self, *args, **kwargs):
+        components = args[0]
+        if isinstance(args[1],(int,float)) and isinstance(args[2],(int,float)) and isinstance(args[3],(int,float)):
+            freq_min = args[1]
+            freq_max = args[2]
+            nfreq = args[3]
+            freq = [0.5*(freq_min+freq_max)] if nfreq == 1 else np.linspace(freq_min,freq_max,nfreq)
+        elif isinstance(args[1],np.ndarray):
+            freq = args[1]
+        else:
+            raise TypeError("add_dft_fields only accepts freq_min,freq_max,nfreq (3 numbers) or freq (numpy array)")
```

Comment:
Rather than have these code copy-and-pasted multiple times, it would be better to have a single function that fixes the argument list.

e.g.
```py
args = fix_freq_args(args, 1)
freq = args[1]
```
where you have something like:
```py
def fix_freq_args(args, i):
     if args.len > i+2 and isinstance(args[i],(int,float)) and isinstance(args[i+1],(int,float)) and isinstance(args[i+2],int):
        freq_min = args[i]
        freq_max = args[i+1]
        nfreq = args[i+2]
        freq = [0.5*(freq_min+freq_max)] if nfreq == 1 else np.linspace(freq_min,freq_max,nfreq)
        del args[i:i+2]
        args[i] = freq
    elseif not isinstance(args[1], (np.ndarray, list)):
        raise TypeError("add_dft functions only accept freq_min,freq_max,nfreq (3 numbers) or freq (array/list)")
```
Note also that `nfreq` has to be an integer and `freq` can be a `list` as well as a numpy array.

Reply:
Unfortunately, this is not possible singe `args` is a tuple which is immutable. We could copy `args` into a list within `fix_freq_args` and return that but that is not so nice. Also, note that because `add_dft_fields` accepts `freq_min`,`freq_max`,`nfreq` (rather than `fcen`,`df`,`nfreq` like all the other `add_dft` functions), it would need to have its own function to set up the `freq` array.

Reply:
Then you can do:
```py
args = fix_freq_args(args, 1)
```py
def fix_freq_args(args, i):
     if args.len > i+2 and isinstance(args[i],(int,float)) and isinstance(args[i+1],(int,float)) and isinstance(args[i+2],int):
        freq_min = args[i]
        freq_max = args[i+1]
        nfreq = args[i+2]
        freq = [0.5*(freq_min+freq_max)] if nfreq == 1 else np.linspace(freq_min,freq_max,nfreq)
        return args[:i] + (freq,) + args[i+3:]
    elseif not isinstance(args[1], (np.ndarray, list)):
        raise TypeError("add_dft functions only accept freq_min,freq_max,nfreq (3 numbers) or freq (array/list)")
    else:
        return args
```

And we should really just change `add_dft_fields`.  It's not acceptable that it is inconsistent with every other DFT function in Meep.

Reply:
The arguments for `add_dft_fields` have been changed from `freq_min`,`freq_max` to `fcen`, `df` to be consistent with the other `add_dft` functions. Some of the `python/tests` which involved `add_dft_fields` have also been updated. (This is only for the Python interface, not Scheme.)

Reply:
Unfortunately, this is a breaking change.   We probably want to emit a BIG WARNING if `df ≥ fcen` in this function for the next few releases to catch people who are using the old API.

Reply:
Done.